### PR TITLE
ARROW-736: [Python] Mixed-type object DataFrame columns should not silently co…

### PR DESCRIPTION
**Pull request by cpcloud**
…erce to an Arrow type by default

### CODE REVIEW

**Comment by wesm** [comment]
needs rebase and cpplint fix

File: cpp/src/arrow/python/pandas_convert.cc [hunk 1]
```diff
@@ -419,21 +420,45 @@ inline Status PandasConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>*
...
+    RETURN_IF_PYERROR();
+  }
+
+  PyObject* bytes_obj = PyUnicode_AsUTF8String(type_name);
+
+  if (bytes_obj == nullptr) {
+    Py_DECREF(type_name);
+    Py_DECREF(type);
+    DCHECK_NE(PyErr_Occurred(), nullptr) << "Call to PyObject_AsUTF8String(...) returned "
+                                            "NULL but PyErr_Occurred() was NULL";
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: Yikes. Perhaps we should develop some tools to help with the Python C API in C++. I have tried to use the OwnedRef to handle RAII-based DECREF

File: cpp/src/arrow/python/pandas_convert.cc [hunk 2]
```diff
@@ -413,6 +418,52 @@ inline Status PandasConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>*
...
+  std::string cpp_type_name(bytes, size);
+
+  std::stringstream ss;
+  ss << "Python object of type " << cpp_type_name << " is not None and is not a "
+     << expected_type_name << " object";
+  auto s = Status::TypeError(ss.str());
+
+  Py_DECREF(bytes_obj);
+  Py_DECREF(type_name);
+  Py_DECREF(type);
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: You could try using the `OwnedRef` for these, save you a bunch of DECREFs per above

File: cpp/src/arrow/python/builtin-convert-test.cc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#include "arrow/array.h"
+#include "arrow/builder.h"
+#include "arrow/python/builtin_convert.h"
+#include "arrow/table.h"
+#include "arrow/test-util.h"
+
+namespace arrow {
+namespace py {
+
+TEST(BuiltinConversionTest, TestMixedTypeFails) {
```
Comment (wesm) [inline cpp/src/arrow/python/builtin-convert-test.cc]: Can you put this in `pandas-test` and rename that test file to `python-test.cc` instead?

File: cpp/src/arrow/python/pandas_convert.cc (hunk 1 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Nice, I'll do that.

File: cpp/src/arrow/python/builtin-convert-test.cc (hunk 3 above)
Comment (cpcloud) [inline cpp/src/arrow/python/builtin-convert-test.cc]: Yep.

**Comment by cpcloud** [comment]
@wesm good to merge? green builds :)

**Review by wesm — APPROVED** [review]
+1, thank you!

**Pull request closed without merging**